lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -1690,12 +1696,17 @@ Type *UpdateFieldTypeForLegacyLayout(Type *Ty,
 
 StructType *UpdateStructTypeForLegacyLayout(StructType *ST,
                                             DxilTypeSystem &TypeSys,
-                                            Module &M) {
+                                            Module &M,
+                                            bool includeTopLevelResource) {
   bool bUpdated = false;
   unsigned fieldsCount = ST->getNumElements();
-  std::vector<Type *> fieldTypes(fieldsCount);
+  std::vector<Type *> fieldTypes;
+  fieldTypes.reserve(fieldsCount);
   DxilStructAnnotation *SA = TypeSys.GetStructAnnotation(ST);
 
+  if (!includeTopLevelResource && dxilutil::IsHLSLResourceType(ST))
```

Comment:
This will be called typically with the element type for the resource, not the resource object type.  Only for StructuredBuffers, or for embedded resources inside another struct (which may be inside cbuffer/tbuffer struct) will this be true.